[WF-117,WF-120] Specify python image as upstream-source + handle failing pebble replan appropriately

shayancanonical opened the pull request:
## Description
<!-- Summary of the changes in this PR -->
- We are not specifying an `upstream-source` for the worker image resource. If a charm specifies an image resource for a workload container, an image must be presented during charm release. The CI to release charm fails since we do not specify one.

To resolve:
1. we specify `python:3.13.7-slim-trixie` as the `upstream-source`. we are unable to use `ubuntu/python` as it is a chiselled image
2. add a check to ensure the charm unit goes into `BlockedStatus` if the workload container does not have the expected entrypoint script

- We are not correctly handling user defined workflows erroring when pebble starts up the worker process. This results in the unit going into `ErrorStatus`, making it a pain to clean up the units.

To resolve:
1. we add exception handling when replanning pebble services to move the unit into `BlockedStatus` to indicate an issue with the provided image. logs may be explored to determine root cause

<!-- Reference the issue this PR addresses (e.g., Closes #123) -->
Closes #69 + Closes #46 
---
## Testing instructions
<!-- Steps to manually test the changes and the expected results, only if applicable -->
<!-- Example:
1. Build charm
4. Configure X with abc value
5. Integrate with Y charm
-->
Added unit tests: `tox -e unit`

## Notes for Reviewers (optional)
<!-- Any specific areas to focus on, known issues, or extra context -->
Manually tested a faulty workload image -> can add an integration test if a good idea

---
## Checklist (all that apply)
- [x] Unit tests added or updated
- [ ] Integration tests added or updated
- [ ] Documentation updated

shayancanonical committed:
Specify python image as upstream-source + handle failing pebble replan appropriately

DnPlas commented on metadata.yaml:
```diff
@@ -45,6 +45,7 @@ resources:
   temporal-worker-image:
     type: oci-image
     description: OCI image containing Temporal worker definition.
+    upstream-source: python:3.13.7-slim-trixie  # use debian based image, ubuntu/python is a chiselled image
```
What's wrong with the chiselled image?

I'd try to avoid something not coming from Canonical, as it would introduce security risks that we'd need to handle eventually.

shayancanonical replied:
A chiselled image is likely functional, but does not include `bash` and so will present challenges to interact with at runtime (we will be unable to attach a shell to the workload container if necessary)

DnPlas replied:
My understanding is that this image's solely purpose is to be a placeholder (and just to fix the release job), users will always have to refresh it with a rock that contains the workflows. If that's the case, then I can't think of a scenario where we'd need a shell. Please let me know if that's true.

shayancanonical replied:
that is correct. i'll update to the `ubuntu/python` stable image

shayancanonical replied:
changed to upstream `ubuntu/python` image in df11d4a99dae40c96a689efcf1f96f8c838d3e0f

DnPlas reviewed (commented):
Thanks @shayancanonical ! Most of the PR looks good to me, just two tiny comments.

Let's also change the sample rockcraft.yaml file. I'd also document this somewhere in the README and update [this docs page](https://charmhub.io/temporal-k8s/docs/t-deploy-worker) (different repo, though).

shayancanonical committed:
Update upstream rock to ubuntu/python image + minor path adjustments in sample rock

shayancanonical committed:
Update resource_sample_py rock README to use absolute paths

shayancanonical commented:
Modified sample rockcraft files to use absolute paths for entrypoint scripts: df11d4a99dae40c96a689efcf1f96f8c838d3e0f
Modified README for resource_sample_py: 3e5d7b8541de7f68cad9cd9543e760f560bf09c8

PR to update docs in temporal-k8s repo + discourse: https://github.com/canonical/temporal-k8s-operator/pull/107

DnPlas reviewed (approved):
Thanks @shayancanonical ! LGTM, the CI is failing, though. Let me know if this needs re-approval after fixing it.
